Fix CoverageSimplifier to handle holes correctly

mukoki opened the pull request:
Fixes #953

dr-jts commented:
Nice fix - thanks.

It would be best to include the two test cases as unit tests.

mukoki commented:
Unfortunately, writing the tests revealed the need to go deeper in the change. Currently, the simplifier removes corners with a distinction between 
- normal edge case (end points are preserved) and 
- ring case (any corner may be removed as far as three distinct points are left).
In the case of a hole touching another hole or touching the shell, we would like to preserve the starting node (which is the node touching the other ring), but edges are passed to the simplifier as a MultiLineString and information that the end node must be preserved is lost. We could preserve initial node everywhen, but for a free ring, it is not required.
Do you agree to pass a bit more information than a MultiLineString to the simplifier ?

dr-jts commented:
> Unfortunately, writing the tests revealed the need to go deeper in the change.

Yay for unit tests!

>   Do you agree to pass a bit more information than a MultiLineString to the simplifier ?

Yes, that seems to be necessary.  What are your ideas for doing this?  Perhaps an array of booleans indicating whether a ring can be fully simplified, or should have its endpoint preserved?

dr-jts commented:
See also #962, which I think requires a different fix.

mukoki commented:
I pushed a new commit adding an attribute to CoverageEdge indicating how many nodes need to be preserved (0, 1 or 2).
It is a much deeper change. Let me know what you think. The main drawback is that I removed public static method from TPVWSimplifier as MultiLineString is not sufficiant to preserve topology :-(
Also added two tests from #953 report

wrt TPVWSimplifier, it could already remove endpoint corner of rings and that was the point ! The LinearRing to CoverageEdge transformation was setting the endpoint on the node touching another ring if any, and but it could be removed as any vertex of a ring. Now it should be preserved (processed as if it was a linestring, not a ring).

See also my comments on #962

dr-jts commented on modules/core/src/main/java/org/locationtech/jts/coverage/CoverageEdge.java:
```diff
@@ -26,12 +26,16 @@ class CoverageEdge {
     Coordinate[] pts = extractEdgePoints(ring, start, end);
-    return new CoverageEdge(pts);
+    CoverageEdge edge = new CoverageEdge(pts);
+    edge.nodeCount = start == end ? 1 : 2;
```
This is only used to check if it is zero, so can it just be a boolean (say `isNodePreserved`) ?

dr-jts commented on modules/core/src/main/java/org/locationtech/jts/coverage/TPVWSimplifier.java:
```diff
@@ -107,13 +115,14 @@ private Geometry simplify() {
-      edges.add(new Edge(line, areaTolerance));
+    for (CoverageEdge covEdge : covEdges) {
+      Edge edge = new Edge(covEdge.getCoordinates(), areaTolerance);
+      edge.nodeCount = covEdge.getNodeCount();
```
Should be passed to constructor?

mukoki replied:
Yes, you're right.

dr-jts commented:
Another alternative is to provide `TPVWSimplifier` with a `Set` of coordinates which are ring nodes that must be preserved.

I prefer this approach, since it reduces the amount of shared classes.  This makes `TPVWSimplifier` more generic.  At some point it may be promoted to being a public class (and moved to the 'simplify` package).

mukoki commented:
I see. In this case, wouln't it be preferable to pass constrained rings (containing one node) as a third MultiLineString beside normal edges (0/2 nodes) and constraints to avoid computation of edges from nodes in TPVWSimplifier.

dr-jts commented:
> Wouln't it be preferable to pass constrained rings (containing one node) as a third MultiLineString beside normal edges (0/2 nodes) and constraints to avoid computation of edges from nodes in TPVWSimplifier.

That is another option.  I think I'd prefer passing the edges which have their endpoints to be preserved (line or ring), and the ones (rings only) which can have all vertices simplified.   Does this make sense?

It's the same amount of work in either case, since the "rings with nodes" have to be determined by checking if their endpoint is in the set of identified nodes.  But maybe it is best to keep this processing outside of `TPVWSimplifier`.

dr-jts commented:
Superseded by #963
